community-code/How_to_Monitor_Diagnose_and_Solve_Gradient_Issues_in_Foundation_Models/gradient_norm_tracking_in_pytorch .ipynb

```diff
@@ -0,0 +1,252 @@
+    "model.to(device)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6",
+   "metadata": {},
+   "source": [
+    "# Step 3. Initialize Neptune for Logging\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "run = Run(\n",
+    "    api_token=\"YOUR_API_TOKEN\",  # replace with your Neptune API token\n",
+    "    project=\"YOUR_WORKSPACE/YOUR_PROJECT\",  # replace with your workspace and project name\n",
+    "    experiment_name=\"gradient_tracking_lr=1\",\n",
+    ")\n",
+    "\n",
+    "run.log_configs(\n",
+    "    {\n",
+    "        \"learning_rate\": 1,\n",
+    "        \"batch_size\": 1,\n",
+    "        \"optimizer\": \"Adam\",\n",
+    "    }\n",
+    ")\n",
+    "run.add_tags([\"gradient_tracking\", \"pytorch\", \"transformers\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8",
+   "metadata": {},
+   "source": [
+    "# Step 4. Define the Gradient Norm Logging Function\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def log_gradient_norms(model, step, log_every_n_steps=1):\n",
+    "    \"\"\"\n",
+    "    Logs L2 norm of gradients for model parameters every n steps using torch.no_grad.\n",
+    "\n",
+    "    Args:\n",
+    "        model (torch.nn.Module): The neural network model.\n",
+    "        step (int): The current training step or epoch, for tracking.\n",
+    "        log_every_n_steps (int): Log only every n steps to reduce overhead.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    if step % log_every_n_steps != 0:\n",
+    "        return  # Skip logging for this step\n",
+    "\n",
+    "    with torch.no_grad():  # Prevent building a computation graph during norm computation\n",
+    "        for name, param in model.named_parameters():\n",
+    "            if param.grad is not None:\n",
+    "                # Optional: skip small/irrelevant layers if needed\n",
+    "                # if not name.startswith(\"encoder.layer.\"): continue\n",
+    "\n",
+    "                grad_norm = param.grad.norm().item()\n",
+    "                run.log_metrics({f\"gradients/{name}\": grad_norm}, step=step)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "10",
+   "metadata": {},
+   "source": [
+    "# Step 5. Train the Model and Track Gradients\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "11",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optimizer = optim.Adam(model.parameters(), lr=1)\n",
+    "\n",
+    "model.train()\n",
+    "for epoch in range(10):\n",
+    "    for step, batch in enumerate(train_dataloader):\n",
+    "        inputs = {k: v.to(device) for k, v in batch.items() if k in tokenizer.model_input_names}\n",
+    "        labels = batch[\"labels\"].to(device)\n",
```

Comment:
I haven't tested it, but it might be faster to move the inputs and labels to the relevant device outside of the loop.